src/PrimeTower.cpp

```diff
@@ -122,42 +166,50 @@ void PrimeTower::generatePaths_denseInfill()
         coord_t current_volume = 0;
         ExtrusionMoves& pattern = pattern_per_extruder[extruder_nr];
 
-        //Create the walls of the prime tower.
+        // Create the walls of the prime tower.
         unsigned int wall_nr = 0;
         for (; current_volume < required_volume; wall_nr++)
         {
-            //Create a new polygon with an offset from the outer polygon.
+            // Create a new polygon with an offset from the outer polygon.
             Polygons polygons = outer_poly.offset(-cumulative_inset - wall_nr * line_width - line_width / 2);
             pattern.polygons.add(polygons);
             current_volume += polygons.polygonLength() * line_width * layer_height * flow;
-            if (polygons.empty()) //Don't continue. We won't ever reach the required volume because it doesn't fit.
+            if (polygons.empty()) // Don't continue. We won't ever reach the required volume because it doesn't fit.
             {
                 break;
             }
         }
 
-        //Only the most inside extruder needs to fill the inside of the prime tower
-        if (extruder_nr != extruder_order.back())
+        // The most outside extruder is used for the base
+        if (extruder_nr == extruder_order.front() && (base_enabled || has_raft) && base_extra_radius > 0 && base_height > 0)
         {
-            pattern_per_extruder_layer0 = pattern_per_extruder;
+            for (coord_t z = 0; z < base_height; z += layer_height)
+            {
+                double brim_radius_factor = std::pow((1.0 - static_cast<double>(z) / base_height), base_curve_magnitude);
```

Comment:
This is a nice implementation because it generically allows other prime tower cross section shapes, but  the only shape fully supported is a circle right?

Reply:
Most of it would work with a different shape indeed, except exclusion computation on the front-end and maybe wipe positions. With some minor effort that should be possible, but I _think_ the PPM team likes the circles because it makes smooth moves.